invites: Do not show streams if user cannot subscribe others.

sahil839 opened the pull request:
We do not show the streams list in invite modal if the user does not have permission to subscribe others to stream.

<!-- Describe your pull request here.-->

 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Screenshot from 2023-05-24 20-14-57](https://github.com/zulip/zulip/assets/35494118/3cedb31f-62f2-4663-b1e5-8c8cd3bfa4aa)

![Screenshot from 2023-05-24 20-16-33](https://github.com/zulip/zulip/assets/35494118/b2adb6b4-1a95-4035-9aed-625f16f3f057)



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
@alya can you please review?

alya commented:
Hm, I feel like the banner will be a bit confusing if you don't know about the feature to configure streams for new users in the first place. Maybe we should do something like this?

> The users you invite will be automatically added to [default streams](/#organization/default-streams-list) for this organization, as you do not have permission to configure which streams new users join.

alya commented:
I'd also just put it at the top where our banners generally go -- I don't think it needs to be in the same place as the stream config.

sahil839 commented:
> Hm, I feel like the banner will be a bit confusing if you don't know about the feature to configure streams for new users in the first place. Maybe we should do something like this?
> 
> > The users you invite will be automatically added to [default streams](/#organization/default-streams-list) for this organization, as you do not have permission to configure which streams new users join.

Will update the banner.

Just one thing to clear, we do not add users to default streams in this case due to recent API changes. Since user is not allowed to add users to the streams, we just do not allow passing anything inside `stream_ids` and thus we consider that as `stream_ids` was empty we do not add users to any streams.

Do we want to change this to set `stream_ids` to default streams if user is not allowed to subscribe others?

alya commented:
> Do we want to change this to set stream_ids to default streams if user is not allowed to subscribe others?

That would make the most sense to me.

sahil839 reviewed (commented):
Updated the PR to have the new message and the message to be on top. Also added commit to subscribe the user to default streams even when the user who invited does not have permission to subscribe others.
![Screenshot from 2023-06-08 20-24-56](https://github.com/zulip/zulip/assets/35494118/7d28500a-4243-4a4c-9d1e-15e5c1ef427e)




Also, added commits to use `subscribe_to_default_streams` parameter such that user is subscribed to default streams at the time of creation of account and not at the time of inviting them. Have posted some questions related to API for Tim.

shameondev commented:
Hey, @sahil839, are there any remaining product questions here or is it just the technical details that need to be discussed?

sahil839 commented:
I don't think so. I have already updated the PR as per Alya's last feedback on the banner and the behavior to subscribe to default streams at the time of actual account creation and not at the time of invitation was also discussed with both Tim and Alya.

I will add the "QA needed" label though, so that this will be tested someone else too.

sahil839 commented:
@mateuszmandera Can you do an initial review of this, especially the backend part?

sahil839 commented on zerver/views/invite.py:
```diff
@@ -83,8 +84,15 @@ def invite_users_backend(
+
+        # We would subscribe the invited user to default streams even when the user
+        # inviting them does not have permission to subscribe others.
+        streams = get_default_streams_for_realm(user_profile.realm_id)
```
I added this as discussed in [this comment](https://github.com/zulip/zulip/pull/25739#issuecomment-1562181618). We do not want to subscribe users to streams when `stream_ids` is passed as empty explicitly. But in case of `not user_profile.can_subscribe_other_users` we do not allow `stream_ids` to be passed as non-empty list and in such case we do not show any UI for selecting streams to users and thus want the invited users to be subscribed to default streams.

timabbott replied:
I guess @mateuszmandera recent work in https://github.com/zulip/zulip/pull/29556 also provides a very clear user case for having it be empty.

sahil839 replied:
And as we also add `subscribe_to_default_streams` parameter, we can probably allow users without permission to subscribe others, to set the parameter as false and not subscribe the new user to default streams. Though as of this PR, we do not allow to user to deselect the `subscribe_to_default_streams` parameter if the user is not allowed to subscribe others to streams.

mateuszmandera reviewed (commented):
Looks reasonable to me  -  didn't look carefully at tests yet though. Posted a couple more comments for now.

sahil839 commented on zerver/views/invite.py:
```diff
@@ -54,6 +53,7 @@ def invite_users_backend(
         default=PreregistrationUser.INVITE_AS["MEMBER"],
     ),
     stream_ids: List[int] = REQ(json_validator=check_list(check_int)),
+    subscribe_to_default_streams: bool = REQ(json_validator=check_bool, default=True),
```
Thinking about it more, I kinda agree that `False` can be a better default. Updated the PR for now. Can change it again if feedback from others suggests otherwise.

Do we want also want to change the DB level default for this then?

timabbott replied:
I think `True` might be fine as a database-level default, but I think we need a `RunPython` migration step to make sure that it's set correctly for preexisting invitation objects -- presumably any that had `stream_ids` set should have a value of `True`. 

While you're it it, we might want to do a multi-step migration where we first add a `null=True` field, then backfill it, and then make it non-null; these can be big tables.

sahil839 commented on zerver/views/registration.py:
```diff
@@ -688,6 +689,10 @@ def prepare_activation_url(
 
+    if subscribe_to_default_streams is not None:
+        prereg_user.subscribe_to_default_streams = subscribe_to_default_streams
+        prereg_user.save()
```
Updated this to call `.save()` only once if at least one of them is not `None`.

sahil839 commented:
Updated the PR. Posted inline comments as well to respond the feedback and mentioning the changes done in brief below.
- Updated the default in API to be `False` as suggested above. Not sure what should be the DB level default.
- Updated the error message for passing both `stream_ids` and `subscribe_to_default_streams` parameter.
- Changed the default value of `subscribe_to_default_streams` to `None` in `accounts_home` in `views/registration.py` which was previously set to `True` since the database level default is already `True` and setting this to `None` prevents an extra call to `.save()`.

alya commented:
Oh yeah, let's do it! @timabbott do you want to take a look now?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11981,6 +11984,17 @@ paths:
+                subscribe_to_default_streams:
+                  description: |
+                    Boolean indicating whether to subscribe the newly created user
+                    to default streams in the organization.
```
Maybe clarify that this is "Boolean indicating that the newly created user should be subscribed to exactly the default streams for the organization.", maybe with "default streams" as a /help/ link.

timabbott replied:
I guess we need to remove the "exactly" language given the new behavior.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -12127,6 +12144,17 @@ paths:
+                    to default streams in the organization.
+
+                    This parameter cannot be passed as `true` if the `stream_ids`
+                    list is not empty.
```
It feels a bit messy to have to mention this every time; has this detail been discussed in the API design stream?

sahil839 replied:
I don't think this has been discussed yet.

timabbott commented on zerver/views/auth.py:
```diff
@@ -272,19 +272,24 @@ def maybe_send_to_registration(
+            prereg_user.subscribe_to_default_streams = subscribe_to_default_streams
         prereg_user.invited_as = invited_as
         prereg_user.multiuse_invite = multiuse_obj
         prereg_user.save()
```
So it looks like one may end up with both `prereg_users.streams` set and `prereg_user.subscribe_to_default_streams` set in this code path?

sahil839 replied:
I guess the `prereg_users.streams` is set only when inviting users either through email or multiuse invites and we have the API restriction to not allow to set both `streams` and `subscribe_to_default_streams`.

But yes, we might want to handle existing `PreregistrationUser` objects since default for `subscribe_to_default_streams` is `True` as you suggested [above](https://github.com/zulip/zulip/pull/25739#discussion_r1580276771).

timabbott commented on zerver/views/invite.py:
```diff
@@ -74,6 +74,9 @@ def invite_users_backend(
     invitee_emails = get_invitee_emails_set(invitee_emails_raw)
 
+    if len(stream_ids) and subscribe_to_default_streams:
+        raise JsonableError(_("Invalid parameters"))
```
This works for me; I think @laurynmm might be working on a very similar string this week related to the stream/channel migration, so ideally we'd make a shared exception class that just takes a list of parameters and uses common code? Not sure.

timabbott commented on zerver/models.py:
```diff
@@ -2367,6 +2367,7 @@ class PreregistrationUser(models.Model):
     full_name_validated = models.BooleanField(default=False)
     referred_by = models.ForeignKey(UserProfile, null=True, on_delete=CASCADE)
     streams = models.ManyToManyField("Stream")
+    subscribe_to_default_streams = models.BooleanField(default=True)
```
`subscribe_to_exactly_default_streams` would be most clear; can you do an "api design" thread about it?

timabbott commented:
This lgtm aside from the handful of detailed noted above; sorry for the slow reviews on this one!

timabbott commented:
> Currently not allowed in the UI, but users who are not allowed to subscribe others can set the `subscribe_to_default_streams` parameter as False. Is this fine?

Probably it's OK to allow it -- since that choice just reduces their initial subscriptions. I suppose it would be more closely tied to letting you remove subscribers, if we tied it to any setting.

timabbott commented:
https://github.com/zulip/zulip/pull/29883 should help with expressing incompatible parameters, but I guess we decided we're not doing that anyway :)

timabbott commented on zerver/actions/create_user.py:
```diff
@@ -139,18 +139,15 @@ def set_up_streams_for_new_human_user(
-            streams = get_slim_realm_default_streams(realm.id)
+        if prereg_user is None or prereg_user.subscribe_to_default_streams:
+            default_streams = get_slim_realm_default_streams(realm.id)
+            streams = list(set(streams) | set(default_streams))
```
My read of the previous logic is that a migration should set:
```
prereg_user.subscribe_to_default_streams = len(prereg_streams.streams) == 0 or (prereg_user.referred_by is None and prereg_user.multiuse_invite is None)
```

to match what the previous logic would have done with a preexisting entry.

timabbott commented:
> Do we still need a migration for existing PreregistrationUser objects to set subscribe_to_default_streams as False where the streams list is non empty? I think we can considering the invitation was originally sent assuming the new user will be subscribed to the streams passed in the list and not the default streams as well.

We should arrange that the behavior for existing invitations is identical to what the behavior would have been before this migration. So we need to match the implicit value of `subscribe_to_default_streams` from the pre-patch code. Posted a comment about what I think this means above. We might also want to avoid setting a default in `do_invite_users` and `do_create_multiuse_invite_link` and instead requiring callers to just pass those parameters -- we better to avoid having duplicate defaults in lots of places.

sahil839 commented:
Added the migration. And I guess the logic should be as below, as there can be case where a user was invited with "streams" parameter set empty explicitly -
```
prereg_user.subscribe_to_default_streams = len(prereg_streams.streams) == 0 and prereg_user.referred_by is None and prereg_user.multiuse_invite is None
```

Also, updated `do_invite_users` and `do_create_multiuse_invite_link` to not have default for `subscribe_to_default_streams`. We can maybe update `do_create_multiuse_invite_link` to require some parameters be mandatory kwargs like we do in `do_invite_users`.

timabbott commented on zerver/migrations/0517_set_subscribe_to_default_streams_for_existing_objects.py:
```diff
@@ -0,0 +1,36 @@
+        subscribe_to_default_streams=False
+    )
+    MultiuseInvite.objects.filter(subscribe_to_default_streams=None).update(
+        subscribe_to_default_streams=False
```
The invitations table is quite large, at the scale of users, but it's not exactly a hot table, so it might be fine to do such a small write to it in a single batch. Would be safer to just batch it and there's no harm in doing so.

timabbott commented:
Logic looks good, let's make the migration safe and then I think we should be able to merge this.

sahil839 commented:
Updated the migration to use batching.

alya commented:
Could you please post an up-to-date screenshot, including the other banners we show at the top now, to make it easy to review how looks all together?

timabbott commented:
I pushed to update the API feature level. I struggled to figure out why the migration query is what it is in this PR --  The core logic change is this:

```
-    user_was_invited = prereg_user is not None and (
-        prereg_user.referred_by is not None or prereg_user.multiuse_invite is not None
-    )
-
     if add_initial_stream_subscriptions:
         # If the Preregistration object didn't explicitly list some streams (it
         # happens when user directly signs up without any invitation), we add the
         # default streams for the realm. Note that we are fine with "slim" Stream
         # objects for calling bulk_add_subscriptions and add_new_user_history,
         # which we verify in StreamSetupTest tests that check query counts.
-        if len(streams) == 0 and not user_was_invited:
-            streams = get_slim_realm_default_streams(realm.id)
+        if prereg_user is None or prereg_user.include_realm_default_subscriptions:
+            default_streams = get_slim_realm_default_streams(realm.id)
+            streams = list(set(streams) | set(default_streams))

```

I feel like there are 3 case:
- User was not invited, but has a `prereg_user` from the signup process; this will have `create_preregistration_user` create one with the Django default of `include_realm_default_subscription=True`. But existing objects will need to be migrated, and these have `referred_by=None, multiuse_invite=None, streams=None`, so that's what your migration targets.
- User was not invited, and is being created by the API, LDAP auth, or some other process that does not generate a preregistration user. In that case, they should get the default streams; that doesn't involve the table though, and the logic change is correct.
- User was invited, in either form. In that case, the previous logic used that list of streams unconditionally, given the `not user_was_invited` check above. So the migration should set `False`, as it does.

OK so I think this is correct ... but let me make that clear in the code comments.

timabbott commented:
OK, added a very detailed docstring for the migration as part of convincing myself that its logic is correct. I renumbered the API documentation and marked this to merge once CI passes, thanks for all the work on this @sahil839!

I also did some copyediting on the docs; one detail worth highlighting is that we should never just say "the user" in API docs; "the current user" or "the target user" is much clearer.

alya commented:
> Could you please post an up-to-date screenshot, including the other banners we show at the top now, to make it easy to review how looks all together?

I guess I can review post-merge.

sahil839 commented:
@alya we do not show the new banner along with existing banners because the existing banners are only shown to admins and admins always have permission to subscribe others. Sharing the updated versions of banner in #30097 which makes the new banner spacing similar to existing banners.

![Screenshot from 2024-05-15 09-09-36](https://github.com/zulip/zulip/assets/35494118/ef8612a8-18e8-417e-a7f3-1594f7b59697)

![Screenshot from 2024-05-15 09-09-14](https://github.com/zulip/zulip/assets/35494118/678e4fe2-e19a-48c8-8381-02f1c69d653c)

alexmv commented on zerver/actions/create_user.py:
```diff
@@ -139,18 +139,15 @@ def set_up_streams_for_new_human_user(
-    )
-
     if add_initial_stream_subscriptions:
         # If the Preregistration object didn't explicitly list some streams (it
         # happens when user directly signs up without any invitation), we add the
         # default streams for the realm. Note that we are fine with "slim" Stream
         # objects for calling bulk_add_subscriptions and add_new_user_history,
         # which we verify in StreamSetupTest tests that check query counts.
```
This comment is now a lie.  @sahil839, can you update it?
